design-documents/services-orchestration.md

```diff
@@ -0,0 +1,254 @@
+              outputPath="params/bar"
+              routine="example:step:1"
+        />
+        <step xsi:type="Task"
+              name="ExecuteStep2"
+              next="ScenarioSuccessfullyCompleted"
+              inputPath="params/bar"
+              routine="example:step:2"
+        />
+        <step name="ScenarioSuccessfullyCompleted" xsi:type="Success" outputPath="params"/>
+    </scenario>
+```
+
+## Routines
+Routines configuration specifies a callable resource to be used in scenario steps.
+
+Actually we have to support two base types of routines:
+* Direct PHP Calls.
+* Magento WebAPI  REST calls.
+
+### Routines declaration example
+```
+    <routine xsi:type="MagentoWebAPIRoutine"
+            name="Magento:Example:Step1"
+            method="PUT"
+            url="http:/magento2sm.local/rest/V1/example/step1"
+    />
+    <routine xsi:type="MagentoWebAPIRoutine"
+             name="Magento:Example:Step2"
+             method="PUT"
+             url="http:/magento2sm.local/rest/V1/example/step2"
+    />
+``` 
+
+[Configuration schema](https://github.com/akaplya/mage-state-machine/blob/master/lib/internal/Magento/Framework/StateMachine/etc/routines.xsd)
+
+## Backward compatibility
+
+Such a mechanism can be useful with the evolution of an existing application.
+Major scenarios can be described with such language.
+For monolithic app the scenario rewrote such way will be backward compatible because scenario will run the same operation in the monolith execution context.
+### Real life example
+
+Implementation of [\Magento\Sales\Api\OrderManagementInterface::place](https://github.com/magento/magento2ce/blob/2.3-develop/app/code/Magento/Sales/Model/Service/OrderService.php#L196) 
+can be represented as the following scenario:
+```xml
+    <scenario name="Magento:Sales:OrderManagement:placeOrder">
+        <step xsi:type="Task" name="PlacePaymentForOrder" next="PersistOrder"
+              inputPath="params/order"
+              outputPath="params/order"
+              routine="Magento:Sales:OrderManagement:placeOrder"
+        >
+            <catch exceptoin="Magento\Payment\Gateway\Command\CommandException" next="PaymentFailureHandler" />
+        </step>
+        <step xsi:type="Task"
+            name="PaymentFailureHandler"
+            inputPath="params/order"
+            exit="true"
+        />
+        <step xsi:type="Task"
+              name="SaveOrder"
+              next="OrderSuccessfullyPlaced"
+              inputPath="params/order"
+              outputPath="params/order"
+              routine="Magento:Sales:OrderManagement:saveOrder"
+        />
+        <step name="OrderSuccessfullyPlaced" xsi:type="Success" outputPath="params/order"/>
+    </scenario>
+```
+For monolith installation this scenario will exactly the same methods with the same arguments as it is now.
+In the same time, such declaration provides new ways how to evolve such service.
+
+## Extensibility
+All steps are declared as XML node. 
```

Comment:
I'm not sure we would be able to refactor all existing scenarios to follow this workflow.